truenas_verify/mtree_verify.py

```diff
@@ -0,0 +1,128 @@
+from collections import namedtuple
+from hashlib import file_digest
+from multiprocessing import Pool
+from os import lstat
+from stat import S_ISDIR, S_ISREG, S_ISLNK, S_IMODE
+import sys
+
+
+LOG_PATH = '/var/log/truenas_verify.log'
+MTREE_FILE_PATH = '/conf/rootfs.mtree'
+MTREE_ENTRY = namedtuple('MtreeEntry', ['fname', 'mode', 'uid', 'gid', 'type', 'link', 'size', 'sha256'])
+
+
+def parse_mtree_entry(line):
+    if line.startswith('#'):
+        return None
+
+    fname, mode, gid, uid, extra = line[1:].split(maxsplit=4)
+    if extra.startswith('type=dir'):
+        entry = MTREE_ENTRY(
+            fname,
+            mode.split('=')[1],
+            int(uid.split('=')[1]),
+            int(gid.split('=')[1]),
+            'dir',
+            None,
+            None,
+            None
+        )
+    elif extra.startswith('type=link'):
+        ftype, link = extra.split()
+        entry = MTREE_ENTRY(
+            fname,
+            mode.split('=')[1],
+            int(uid.split('=')[1]),
+            int(gid.split('=')[1]),
+            'link',
+            link.split('=')[1],
+            None,
+            None
+        )
+    else:
+        ftype, size, shasum = extra.split()
+        entry = MTREE_ENTRY(
+            fname,
+            mode.split('=')[1],
+            int(uid.split('=')[1]),
+            int(gid.split('=')[1]),
+            ftype.split('=')[1],
+            None,
+            int(size.split('=')[1]),
+            shasum.split('=')[1]
+        )
+
+    return entry
+
+
+def validate_file_sha256sum(entry, errors):
+    with open(entry.fname, 'rb', buffering=0) as f:
+        hash = file_digest(f, 'sha256').hexdigest()
+        if hash != entry.sha256:
+            errors.append(f'{entry.fname}: expected: {entry.sha256}, got: {hash}')
+
+
+def validate_mtree_entry(entry) -> list[str]:
+    try:
+        st = lstat(entry.fname)
+    except FileNotFoundError:
+        return [f'{entry.fname}: file does not exist.']
+
+    assert st.st_uid == entry.uid, entry.fname
```

Comment:
Should these be appended to error list rather than simply asserting?

Reply:
Yeah I'll change this